src/sknnr_spatial/image/_base.py

```diff
@@ -0,0 +1,193 @@
+    def _flatten(self, image: ImageType) -> ImageType:
+        """
+        Ravel the image's x, y dimensions while keeping the band dimension.
+        """
+
+    @abstractmethod
+    def unflatten(self, flat_image: ImageType, *, apply_mask: bool = True) -> ImageType:
+        """
+        Reconstruct the x, y dimensions of a flattened image to the original shape.
+        """
+
+    def _get_nodata_mask(self, flat_image: ImageType) -> ImageType | None:
+        """
+        Get a mask of NoData values in the shape (pixels,) for the flat image.
+
+        NoData values are represented by True in the output array.
+        """
+        # Skip allocating a mask if the image is float and nodata wasn't given
+        if not (is_float := flat_image.dtype.kind == "f") and self.nodata_vals is None:
+            return None
+
+        mask = self._backend.zeros(flat_image.shape, dtype=bool)
+
+        # If it's floating point, always mask NaNs
+        if is_float:
+            mask |= self._backend.isnan(flat_image)
+
+        # If nodata was specified, mask those values
+        if self.nodata_vals is not None:
+            mask |= flat_image == self.nodata_vals
+
+        # Set the mask where any band contains nodata
+        return mask.max(axis=self.flat_band_dim)
+
+    def _validate_nodata_vals(
+        self, nodata_vals: float | tuple[float] | NDArray | None
+    ) -> NDArray | None:
+        """
+        Get an array of NoData values in the shape (bands,) based on user input.
+
+        Scalars are broadcast to all bands while sequences are checked against the
+        number of bands and cast to ndarrays. There is no need to specify np.nan as a
+        NoData value because it will be masked automatically for floating point images.
+        """
+        if nodata_vals is None:
+            return None
+
+        # If it's a numeric scalar, broadcast it to all bands
+        if isinstance(nodata_vals, (float, int)) and not isinstance(nodata_vals, bool):
+            return np.full((self.n_bands,), nodata_vals)
+
+        # If it's not a scalar, it must be an interable
+        if not isinstance(nodata_vals, Sized) or isinstance(nodata_vals, (str, dict)):
+            raise TypeError(
+                f"Invalid type `{type(nodata_vals).__name__}` for `nodata_vals`. "
+                "Provide a single number to apply to all bands, a sequence of numbers, "
+                "or None."
+            )
+
+        # If it's an iterable, it must contain one element per band
+        if len(nodata_vals) != self.n_bands:
+            raise ValueError(
+                f"Expected {self.n_bands} nodata values but got {len(nodata_vals)}. "
+                f"The length of `nodata_vals` must match the number of bands."
+            )
+
+        return np.asarray(nodata_vals, dtype=float)
+
+    def _fill_nodata(self, flat_image: ImageType, nodata_fill_value=0.0) -> ImageType:
+        """Fill values in a flat image based on the pre-calculated mask."""
```

Comment:
Alas, just me being inconsistent! The [numpydoc style guide](https://numpydoc.readthedocs.io/en/latest/format.html) seems to suggest using this for one-line comments:

```python
"""Docstring style 1

"""
``` 

But poking through the Numpy source, they seem to use a mix of formats including 

```python
"""Docstring style 2"""
```

and 

```python
"""
Docstring style 3
"""
```

I think they'll all parse fine, so probably just a question of preference. Personally I'm not a fan of the recommended "style 1" since it wastes more vertical room than "style 2" and more horizontal room than "style 3". It looks like I *mostly* used "style 2" here, so I'll (try) to stick to that from now on unless you've got strong feelings on the matter.

Reply:
Just came across this in [PEP 8](https://peps.python.org/pep-0008/) while addressing another comment:

> For one liner docstrings, please keep the closing """ on the same line:
"""Return an ex-parrot."""

Reply:
> It looks like I _mostly_ used "style 2" here, so I'll (try) to stick to that from now on unless you've got strong feelings on the matter.

Happy to follow your lead here.  I've personally always used "style 3", but it adds a lot of unneeded whitespace.  And just to be fully pedantic, would a multi-line docstring always be:
```
"""Docstring for a multi-line with lots of lots and lots of words and some
bells and whistles."""
```
or do you change to numpy style or even "style 3"?  Looks like we've used a mix in `sknnr`.

Reply:
I couldn't find a clear answer to that in PEP 8 or PEP 257. They both seem to assume that a docstring is either a single sentence that fits on one line ("style 2") or multiple sentences that are split over multiple lines ("style 3") and don't cover the case where a single sentence can't fit on a single line. I guess the implication is probably to avoid that whenever possible in favor of a shorter description with additional details below, but if it was unavoidable I'd lean towards "style 3". In other words, I would propose:

Single-line:
```python
"""Fits on one line."""
```

Single-line that's too long (avoid when possible):
```python
"""
Doesn't fit on one line because it has too many words that were all very important
to include.
"""
```


Multi-line:
```python
"""
Short summary that fits on one line.

Additional details below.
"""
```

With that said, I'll probably forget this as soon as I write my next docstring, so please call me out if I'm being inconsistent!